### ISSUE TITLE
How to get Variable Value from Host to Service Variable

### ISSUE THREAD

**Issue by firatalkis** [Icinga/icingaweb2-module-director#1965]
Hi,

We have installed Icinga Directory Module. We want to get Host Variable value and use this value into the Service. But even if I used the "$host.vars.enviroment$" variable, I couldn't get any result. And also I tried below variables as well;
$vars.enviroment$, $host_vars_enviroment$, host_vars_enviroment etc.

This variable type is string.

When we used these variables, the results are like below screenshots.

![image](https://user-images.githubusercontent.com/42132281/65860485-171c8500-e373-11e9-8cce-e9ee93d3772b.png)

Could you please help us on this issue ?

Tnx.

## Your Environment

* Director version (System - About): 1.6.2
* Icinga Web 2 version and modules (System - About): 2.7.0
* Icinga 2 version (`icinga2 --version`): 2.10.5
* Operating System and version: Red Hat Linux 7.6
* Webserver, PHP versions: 7.1

**Comment by Thomas-Gelf** [comment]
Reading [this](https://icinga.com/docs/icinga2/latest/doc/03-monitoring-basics/#evaluation-order) might help. Or eventually ask for help in our [community forum](https://community.icinga.com), this tracker here is for bugs and feature requests.

**Comment by firatalkis** [comment]
We have not integrated the director module in our current Icinga2 application and when we want custom host variable value or custom service variable value, we could getting it's value over service object, notification object hostgroup object, servicegroup object etc. so we could getting it's value over the all objects.(host.name, host.vars.environment, "$host.vars.domain$", service.name, "$service.vars.disk_crit$" etc)

When we tried the director module (v1.6.2) in our test environment, this feature didn't work in the director module, and that's a shortcoming and bug for us. We used $host.vars.environment$ as a spelling and tried many different methods ($host_vars_environment$, $vars.environment$, etc.).

Could the director module is not object-oriented?

and

We wrote this case at community forum but we don't find any solution. You can see in following link.

[https://community.icinga.com/t/getting-value-for-service-from-host/1890](url)

and

When we met Blerim Sheqa in Icinga Camp Milan, we spoke this case. Then he suggested that we write this case to github.

Thanks.

**Comment by Thomas-Gelf** [comment]
@firatalkis, did you read the linked documentation? And people in the forum also pointed you to the right direction. Your host's `environment` variable is always available on it's services, and it's commands.

**Comment by firatalkis** [comment]
@Thomas-Gelf,  Yes i read linked documentation. So why not get the value even though I use a this format $host.vars.environment$ in service object at director module.

**Service Properties**
![image](https://user-images.githubusercontent.com/42132281/66812824-4bc83900-ef3c-11e9-8a6e-393fc8583e1f.png)

**Host Properties**
![image](https://user-images.githubusercontent.com/42132281/66813211-09ebc280-ef3d-11e9-83d4-62c9b94ecd4a.png)

and i look at the object inspection but i don't see the hots environment value
![image](https://user-images.githubusercontent.com/42132281/66813755-06a50680-ef3e-11e9-98c7-1055a8cda7bb.png)


Where am i doing wrong?

**Comment by stevie-sy** [comment]
For @Thomas-Gelf:
I tried this also and it doesn't work. If I understand @firatalkis after the last screenshots right, he wants to refer the value of a variable from the service to a variable from the host.  In this should be displayed in icingaweb2.
My colleague found an old thread from the monitoring-portal where @dnsmichi described why this is not possible: https://monitoring-portal.org/woltlab/index.php?thread/31327-icinga2-host-variablen-in-service-nutzen/. Maybe that brings some light into this.

**Comment by Thomas-Gelf** [comment]
@stevie-sy: that's not true, it just works. Just different than one might expect. Above example renders as follows:

```
object Service "some-service" {
  host_name = "my-host"
  vars.something = "$host.vars.something$"
}
```

Icinga interprets this as:

```
object Service "some-service" {
  host_name = "my-host"
  vars.something = macro("$host.vars.something$")
}
```

So it never has a property, and that's why it is not written to the IDO. It's a method returning another property every time you read from it. If you want to have it "constant", you need to write as follows:

```
object Service "some-service" {
  host_name = "my-host"
  vars.something = host.vars.something
}
```

BUT this only works for Apply-Rules, the host-Property isn't available in Service objects. Please don't ask me why, I had this discussion many times before - and got no satisfying answer. So, to get something similar to work you would have to render it as follows:

```
object Service "some-service" {
  host_name = "my-host"
  vars.something = get_host("my_host").vars.something
}
```

It get's even more complicated if you want to place this into a template, because you could use the template for Apply Rules AND for single Service objects. Of course all this is possible, and the Director uses such constructs since quite some time. Please check `DirectorOverrideTemplate` for a working example - it is rendered with every deployment to simplify some tricky tasks.

Long story short: the Director rendered what @firatalkis told him to render. He doesn't "see" the value in the Icinga Web 2 monitoring module, as Icinga is unable to write a function/pointer to the database. But the value is still there. And it is completely useless anyways, as you can access Host properties from every command, every Apply Rule - no need to clone it to every service.

Hope this helps.

**Comment by firatalkis** [comment]
Thanks @stevie-sy
As i understand it this macros works in runtime but not display on icingaweb2. Because the director module don't keep custom variable values in own db.

We see the host's custom variable value on the host service because we don't use director module yet. Icinga2 having at the own database custom variable values.

Thanks a lot for your answers.

**Comment by stevie-sy** [comment]
Thanks @Thomas-Gelf . I think I put it wrong. It doesnt'work like @firatalkis -> Display the value from the host variable in the service variable in icingaweb2. 
We use this way like you described. So that means we define some thresholds at the host definition. And after we refer this values to a variable at the Service. And yes it is interpreted as a macro. icingaweb2 display this also that way. For us it's fine. And that behaviour you decribed is also described in the thread from Monitoring-Portal I linked. 
But for me it was only a guess what he could mean, after reading his issue and his thread at community.icinga.com. I just wanted to help ;-) 

btw: thanks for the hint with 
`vars.something = get_host("my_host").vars.something
`
I didn't knew that. (y)

**Comment by firatalkis** [comment]
Thanks @stevie-sy, thanks @Thomas-Gelf 
As i understand it this macros works in runtime but not display on icingaweb2. Because the director module don't keep at the database custom variable values 

We see the host's custom variable value on the host service because we don't use director module yet. Icinga2 having at the own database custom variable values.

Thanks a lot for your answers.

**Comment by Thomas-Gelf** [comment]
@firatalkis: that's correct, apart from the DB. Director isn't in the pipeline there, and it doesn't matter whether or not you're using it. Icinga 2 is writing to the Icinga IDO database, and the `monitoring` module reads from that database.

May I ask what your use-case is, why are you trying to replicate a host property onto each of it's services?

**Comment by stevie-sy** [comment]
In our case:
We use different thresholds e.g. for the disk-checks. or exclude drives to check from windows-Servers by the icinga Agent.
Since our director workshop we know, that we could import services or we use the DirectorOverride-function.
But we Import the informations from our central CMDB and other sources. And we wanted the information in a central position. So a good way was to use the custom vars from every host. And with that Information we manage our ServiceSets and ApplyRules. We use them a lot.

**Comment by firatalkis** [comment]
it's enough for us to actually work in the back.

The reason I want to do this is because the metrics collected by Icinga2 are in the host information in the series generated when sending to influxdb.

**Comment by Thomas-Gelf** [comment]
@stevie-sy: in your scenario I'd strongly advice for and Import/Sync-Rule generating/purging single Services according your CMDB. But glad to hear that it works anyways :)
@firatalkis: perfect, then you should be covered
